src/test/java/com/gildedrose/GildedRoseTest.java

```diff
@@ -2,16 +2,160 @@
+        app = new GildedRose(items);
+    }
+
+    @Test
+    public void doesNotAlterName() {
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_NAME, app.items[0].name);
+    }
+
+    @Test
+    public void doesDecreaseSellIn() {
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_SELLIN - 1, app.items[0].sellIn);
+    }
+
+    @Test
+    public void doesDecreaseQuality() {
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_QUALITY - 1, app.items[0].quality);
+    }
+
+    @Test
+    public void withPassedSellIn_doesDecreaseQuality() {
+        item.sellIn = ITEM_OVERDUE_SELLIN;
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_QUALITY - 2, app.items[0].quality);
+    }
+
+    @Test
+    public void doesNotDecreaseQualityBelowZero() {
+        item.quality = ITEM_QUALITY_MIN;
+        app.updateQuality();
+        assertEquals(ITEM_QUALITY_MIN, app.items[0].quality);
+    }
+
+    @Test
+    public void withNegativeQuality_doesNotDecreaseQuality() {
+        item.quality = -1;
+        app.updateQuality();
+        assertEquals(-1, app.items[0].quality);
+    }
+
+    @Test
+    public void withAgedBrie_doesIncreaseQuality() {
+        item.name = ITEM_NAME_AGED_BRIE;
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_QUALITY + 1, app.items[0].quality);
+    }
+
+    @Test
+    public void withAgedBrieAfterSellIn_doesIncreaseQualityTwice() {
+        item.name = ITEM_NAME_AGED_BRIE;
+        item.sellIn = ITEM_OVERDUE_SELLIN;
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_QUALITY + 2, app.items[0].quality);
+    }
+
     @Test
-    public void foo() {
-        Item[] items = new Item[] { new Item("foo", 0, 0) };
-        GildedRose app = new GildedRose(items);
+    public void withLongOverdueAgedBrie_doesIncreaseQualityTwice() {
+        item.name = ITEM_NAME_AGED_BRIE;
+        item.sellIn = ITEM_LONG_OVERDUE_SELLIN;
         app.updateQuality();
-        assertEquals("fixme", app.items[0].name);
+        assertEquals(ITEM_DEFAULT_QUALITY + 2, app.items[0].quality);
     }
 
+    @Test
+    public void withAgedBrie_doesNotIncreaseQualityOverMaximum() {
+        item.name = ITEM_NAME_AGED_BRIE;
+        item.quality = ITEM_QUALITY_MAX;
+        app.updateQuality();
+        assertEquals(ITEM_QUALITY_MAX, app.items[0].quality);
+    }
+
+    @Test
+    public void withSulfuras_doesNotChangeQuality() {
+        item.name = ITEM_NAME_SULFURAS;
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_QUALITY, app.items[0].quality);
+    }
+
+    @Test
+    public void withSulfuras_doesNotChangeSellIn() {
+        item.name = ITEM_NAME_SULFURAS;
+        app.updateQuality();
+        assertEquals(ITEM_DEFAULT_SELLIN, app.items[0].sellIn);
+    }
+
+    @Test
+    public void withBackstagePassToBeSoldInElevenDays_increasesQualityByOne() {
+        item.name = ITEM_NAME_BACKSTAGE_PASS;
+        item.sellIn = 11;
```

Comment:
could have extracted constants defining the limits of the various "life phases" of a backstage pass.